hugr-py/src/hugr/hugr.py

```diff
@@ -0,0 +1,313 @@
+        else:
+            node = Node(len(self._nodes))
+            self._nodes.append(node_data)
+        return node
+
+    def delete_node(self, node: Node) -> NodeData | None:
+        for offset in range(self.num_in_ports(node)):
+            self._links.delete_right(node.inp(offset))
+        for offset in range(self.num_out_ports(node)):
+            self._links.delete_left(node.out(offset))
+
+        weight, self._nodes[node.idx] = self._nodes[node.idx], None
+        self._free_nodes.append(node)
+        return weight
+
+    def add_link(self, src: OutPort, dst: InPort) -> None:
+        src = _unused_sub_offset(src, self._links)
+        dst = _unused_sub_offset(dst, self._links)
+        if self._links.get_left(dst) is not None:
+            dst = replace(dst, sub_offset=dst.sub_offset + 1)
+        self._links.insert_left(src, dst)
+
+    def delete_link(self, src: OutPort, dst: InPort) -> None:
+        self._links.delete_left(src)
+
+    def num_nodes(self) -> int:
+        return len(self._nodes) - len(self._free_nodes)
+
+    def num_in_ports(self, node: Node) -> int:
+        return len(self.in_ports(node))
+
+    def num_out_ports(self, node: Node) -> int:
+        return len(self.out_ports(node))
+
+    def in_ports(self, node: Node) -> Collection[InPort]:
+        # can be optimised by caching number of ports
+        # or by guaranteeing that all ports are contiguous
+        return [p for p in self._links.bck if p.node == node]
+
+    def out_ports(self, node: Node) -> Collection[OutPort]:
+        return [p for p in self._links.fwd if p.node == node]
+
+    def insert_hugr(self, hugr: "Hugr", parent: Node | None = None) -> dict[Node, Node]:
+        mapping: dict[Node, Node] = {}
+
+        for idx, node_data in enumerate(hugr._nodes):
+            if node_data is not None:
+                mapping[Node(idx)] = self.add_node(node_data.op, node_data.parent)
+
+        for new_node in mapping.values():
+            # update mapped parent
+            node_data = self[new_node]
+            node_data.parent = mapping[node_data.parent] if node_data.parent else parent
+
+        for src, dst in hugr._links.items():
+            self.add_link(
+                mapping[src.node].out(src.offset), mapping[dst.node].inp(dst.offset)
+            )
+        return mapping
+
+    def to_serial(self) -> SerialHugr:
+        node_it = (node for node in self._nodes if node is not None)
+        return SerialHugr(
+            version="v1",
+            # non contiguous indices will be erased
+            nodes=[node.to_serial(Node(idx), self) for idx, node in enumerate(node_it)],
+            edges=[
+                ((src.node.idx, src.offset), (dst.node.idx, dst.offset))
+                for src, dst in self._links.items()
+            ],
+        )
+
+    @classmethod
+    def from_serial(cls, serial: SerialHugr) -> "Hugr":
+        raise NotImplementedError
+
+
+@dataclass()
+class Dfg:
```

Comment:
Should we think here about a commands-based DFG construction similar to tket's?
(E.g. `dfg.add(CNot(target=a, control=b))`)

Reply:
see https://github.com/CQCL/hugr/pull/1098/commits/de22b5a646046296d6c7080819e7eceff69fad4d...